### DISCUSSION TITLE
set FPS

### DISCUSSION THREAD

**Discussion by AldoReset** [hajimehoshi/ebiten#1563] [Q&A]
Is seems not possible to change fps. it would be interesting to fix max value.

for example monitor and card may do : 1/60 , 1/100 or 1/120 but you can fix to 1/50 fps.
and also have a func can return vbl and hbl when possible.

**Comment by hajimehoshi** [comment]
FPS is determined by a player's environment like the display's refresh rate. This is not under control.

Instead you can modify TPS (tick-per-second) by `SetMaxTPS`.

**Reply by hajimehoshi** [reply]
FPS is a frequency how often `Draw` is called, while TPS is a frequency how often `Update` is called. Ebiten can control TPS. Does this make sense?

**Reply by Zyko0** [reply]
Some games allow for a parameter to cap FPS to a maximum value (vsync being on or off), maybe that's what OP refers to.
For some reason even tho your gpu + the game itself lets you reach 240 constant, you might want to set the maximum to 120-140-160, it is an option available to some existing games.

edit: We could imagine it the same way Ebiten actually builds `updateCount` with `TPS`, but a `drawCount` with `maxFPS` for a start

**Comment by erexo** [comment]
Let me dig this discussion up and try to ask for the `SetMaxFPS` API call once again.
Some of my players are playing on a very old piece of hardware where every ounce of performance matters. Games can squeeze every last breath of processing power just to reach that 60fps and sure, it may work and players might be able to play smoothly but the performance of other running applications might suffer, so rarely players decide to cap their FPS to 30-40 just to be able to use their computers while the game is running.
It might or might not work with the Vsync, ie. vsync might be a cap for max FPS just like now, but with additional, optional constant integer that could be determined programmatically.
Please consider this once again, thanks.

**Reply by hajimehoshi** [reply]
IIUC a display's refresh rate cannot be changed dynamically. How can we adjust FPS?

**Reply by Zyko0** [reply]
Maybe there is a way in ebiten's logic to do something with timers, and skip draw calls (not clearing screen) until it's time to draw (with some exceptions that require redraw, like window resizing etc..)?

**Reply by hajimehoshi** [reply]
Hm, so we can emulate such skipping without new APIs, right? (This doesn't mean I will never introduce a new API, but I want to confirm what we actually need)

**Reply by hajimehoshi** [reply]
One issue is that there is no portable way to get the current refresh rate (especially on browsers)

**Reply by Zyko0** [reply]
> Hm, so we can emulate such skipping without new APIs, right?

We would require a new method to set the desired max refresh rate I believe

> One issue is that there is no portable way to get the current refresh rate (especially on browsers)

Can we just manage it internally? Let's say the user ask for 40fps, ebiten would only call `Draw()` every 1/40
Of course, this could probably only work with vsync off in order to respect a consistent interval (and it would also allow to cap at 360 for example)

**Comment by tardisx** [comment]
I have a use case (which might be silly or have another approach - I am new to this and suggestions are welcome).

I am implementing a multiplayer board game, and most of the time the display will be static, or nearly static. I am burning a lot of CPU of my users for no reason (WASM, in browser context).

30fps would presumably halve the cost, and for a board game, no one would mind :-)

Even better would be to dynamically adjust it, if I know something is about to happen, I can increase FPS for smooth animations, then turn it down again.

**Reply by hajimehoshi** [reply]
As of v2.5, if you call `SetScreenClearedEveryFrame(false)` and don't touch the screen at `Draw`, Ebitengine optimizes GPU usages so you don't have to care about adjsting FPS for relatively static applications.

**Reply by tardisx** [reply]
That will be amazing! Thank you for this and all of ebitengine @hajimehoshi !

**Reply by erexo** [reply]
Remember to use `SetScreenClearedEveryFrame(false)` only with vsync on, otherwise fps may skyrocket and the cpu consumption will be a lot bigger
